enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,213 @@
+---
+title: microshift-pluggable-core-components
+authors:
+  - majopela
+reviewers:
+  - "@copejon, MicroShift contributor"
+  - "@fzdarsky, MicroShift architect"
+  - "@ggiguash, MicroShift contributor"
+  - "@dhellmann, MicroShift contributor"
+  - "@oglok, MicroShift contributor"
+  - "@zshi-redhat, MicroShift "
+approvers:
+  - "@dhellmann"
+api-approvers:
+  - None
+creation-date: 2022-12-19
+last-updated: 2022-12-19
+tracking-link:
+  - https://issues.redhat.com/browse/USHIFT-599
+---
+
+# MicroShift Pluggable Core Components
+
+## Summary
+
+This enhancement proposes making some of the core components like CNI, CSI or router
+optional and pluggable.
+
+In the enhancement proposal we will describe an architecture which would allow
+customers to install individual compomnents as rpm packages.
+
+
+## Motivation
+
+The principal factor that gave birth to the idea of MicroShift was the need to bring
+the OpenShift deployment model into the Edge where the power, memory and CPU budgets
+are reduced.
+
+While the default configuration makes sense to most of our customers in the Device
+Edge area, still our current CSI, CNI, and router components use 35% of the memory
+footprint and a significative CPU and power footprint. Making those components
+pluggable would allow customers who don't require advanced network features,
+persistent volumes, or ingress/routes to disable or switch those technologies
+and reduce the memory/cpu and disk/over-the-wire footprint.
+
+
+### User Stories
+
+An application developer needs the whole stack of components, creates an image
+that contains:
+1. microshift
+
+Nothing would change in this case, installing microshift, which becomnes a virtual
+package, installs the necessary dependencies:
+* microshift-core
+* microshift-networking-ovn
+* microshift-storage-topolvm
+* microshift-routes
+
+as well as any other dependencies.
+
+An application developer only needs advanced networking, but no persistent volumes,
+and no routes/ingress.  Creates an image that contains:
+1. microshift-core
+2. microshift-networking-ovn
+
+
+An application developer wants the bare minimum set of components creates an image
+with:
+1. microshift-core
+
+An application developer wanting bare minimum networking but persistent volumes
+creates an image with:
+1. microshift-core
+2. microshift-storage-topolvm
+
+### Goals
+* Allow networking, storage and router core components to be optionally
+  installed.
```

Comment:
We also need to make CNI/CSI replaceable for upstream work.

Reply:
Ack, I will make sure this is properly handled in the goals and work items.

Reply:
Ah, in that case we would need etc dir support, but that would mean that the config management needs to take care of versioning sync with microshift when necessary, not sure how I feel about it.